### PR TITLE
Refactor/combine RED.waitingYielders()/RED.waitingForSignals() into a…

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
… single method

This refactors/combine the
RequestEventDispatcher.waitingYielders()/waitingForSignals() into a
single method, using a single pass over the waiting fibers array to
set the fields.

### CODE REVIEW

**Comment by nemanja-boric-sociomantic** [comment]
The diff is a bit messed up, but it's a simple change - remove `waitingForSignal`/`waitingYielders.` and replace them with the single method with suboptimal name.

File: src/swarm/neo/request/RequestEventDispatcher.d [hunk 1]
```diff
@@ -761,35 +757,39 @@ public struct RequestEventDispatcher
...
-                waiting_fiber.event.active == waiting_fiber.event.active.signal )
-                return true;
+        {
+            if ( waiting_fiber.enabled )
+            {
+                switch (waiting_fiber.event.active)
+                {
+                    case waiting_fiber.event.active.signal:
+                        flags |= flags.Resume;
+                        if (flags & ~cast(int)(flags.Resume))
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/request/RequestEventDispatcher.d]: Hmm... wouldn't `flags.Receive` be required to also be set for this condition to be true?

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/request/RequestEventDispatcher.d]: This just does early return if the `Yield` flag was already set, so we can stop the iteration. Unless I'm missing something? I'm just comparing to non-zero of the resulting bitwise and.

So, if there was a yield set previously:

```
[resume][yield][receive]
0       1       0
```

and-ing this with `011` (`~resume`) will yield `!= 0` so we can return immediately  .

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/request/RequestEventDispatcher.d]: Ahh gotcha. Would `if ( flags & flags.Yield )` be clearer and equivalent? (As we're only talking about yield / resume.)

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/request/RequestEventDispatcher.d]: Yeah, that's the way I originally wrote this, but this is the more friendly when it comes to refactoring and adding the next code (in each case you just repeat the flag you're setting).

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/request/RequestEventDispatcher.d]: Oh, no, that's wrong!

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/request/RequestEventDispatcher.d]: Hmmm... it's tricky though. You only want to return if all flags _that this function deals with_ are set. (i.e. not `Receive`.)

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/request/RequestEventDispatcher.d]: Updated.

File: src/swarm/neo/request/RequestEventDispatcher.d [hunk 2]
```diff
@@ -761,35 +757,42 @@ public struct RequestEventDispatcher
...
+                        break;
+                    case waiting_fiber.event.active.yield:
+                        flags |= flags.Yield;
+                        break;
+                    default:
+                        break;
+                }
 
-        return false;
+                if (flags == all_set_flags)
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/request/RequestEventDispatcher.d:790]: Much nicer :+1:

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Pull request merged**